proposals/add-array-object-feature-option-types.md

```diff
@@ -0,0 +1,170 @@
+# Proposal: Add `array` and `object` Types to Feature Options Schema
+
+## Problem Statement
+
+The current Dev Container Feature Options Schema lacks support for `array` and `object` types. This limitation hinders the ability to configure features that require lists of values (e.g., dependencies, ports) or key-value mappings (e.g., environment variables, custom settings). Addressing this limitation will enhance the flexibility and usability of the schema.
+
+This proposal aims to address [Issue #567](https://github.com/devcontainers/spec/issues/567).
+
+## Proposed Solution
+
+Extend the Dev Container Feature Options Schema by introducing `array` and `object` as valid types for feature options. These new types will enable developers to create more complex and customizable features.
+
+### Schema Changes
+
+1. **Array Type**:
+   - Represents a list of values (e.g., strings, numbers, booleans, or objects).
+   - Supports optional constraints such as allowed item types and default values.
+
+2. **Object Type**:
+   - Represents a mapping of key-value pairs.
+   - Allows nested structures and supports default values for specific keys.
+
+### Example Feature Using `array` and `object`
+
+Below are two examples of how `array` and `object` types can be used in feature configurations.
+
+#### Example 1: Environment Variables
+
+**Feature Definition (`example-feature/devcontainer-feature.json`):**
+
+```json
+{
+  "name": "example-feature",
+  "id": "example",
+  "description": "A feature that demonstrates the use of array and object types.",
+  "options": {
+    "dependencies": {
+      "type": "array",
+      "description": "List of software packages to install",
+      "default": ["git", "curl", "nodejs"]
+    },
+    "envVariables": {
+      "type": "object",
+      "description": "Environment variables for the container",
+      "default": {
+        "NODE_ENV": "development",
+        "APP_PORT": "3000"
+      }
+    }
+  },
+  "install": {
+    "script": "./install.sh",
+    "args": []
+  }
+}
+```
+
+**Installation Script (`example-feature/install.sh`):**
+
+```bash
+#!/bin/bash
+
+# Install dependencies
+echo "Installing dependencies..."
+for package in "${DEPENDENCIES[@]}"; do
```

Comment:
This seems to be Bash specific syntax. How would this work with the Bourne shell `/bin/sh`? The Bourne shell is used for the install scripts to allow for compatibility with distributions not including Bash by default like Alpine Linux.